### DISCUSSION TITLE
Annotations in multi-layer files

### DISCUSSION THREAD

**Discussion by djoltes** [pymupdf/PyMuPDF#2746] [Looking for help]
I'm still working on my redaction solution and discovered that, interestingly, if I add an annotation to a document containing OCR using typical methods (as below) I end up with 2 highlights in the same location. It appears that each layer is getting its own highlight, so if you (for example) open a file containing these highlights you have to delete the highlight twice for it to disappear.

![image](https://github.com/pymupdf/PyMuPDF/assets/11085091/befce756-6d5c-4cb4-af5e-aa6aabf4045b)

Is there a mechanism for applying an annotation to only one "layer" of a PDF?

**Comment by JorjMcKie** [comment]
By "layer" you mean the image and the OCRed text, right? Not the same-named Optional Content concept.

No:
If you add an annotation once, it will be there only one time. Annotations know the heck about OCR, image, text or whatever stuff on the page.
Best way understanding this is the following metaphor:
* the PDF page is like a work of art, a picture in a frame with a protecting glass above it
* an annotation is something painted on that glass: it has connection at all to the picture underneath and can be wiped away without inflicting the picture.

So if you have two annotations in the same place, then you must have inserted two.

**Reply by djoltes** [reply]
Okay, that's what I suspected so there must be something in the code at my end. All I'm doing is iterating over what Comprehend has returned, so unless it's somehow flagging the same PII >1 time it must be my issue. Thanks!

**Reply by JorjMcKie** [reply]
You you store the rectangle already "served" with an annot and introduce an existence check.

**Reply by djoltes** [reply]
I considered that, but Comprehend is reporting the same content multiple times with slightly different locations, which seems really odd.  That's making it really difficult to build a list of previously visited elements, though I suppose I could just use the text.

![image](https://github.com/pymupdf/PyMuPDF/assets/11085091/9c91ae50-e7a2-4142-8c3d-80d3b85a3c5f)
